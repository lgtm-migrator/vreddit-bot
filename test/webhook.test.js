const { mocked, CHAT, withFnMocks, ctx, FROM, fnMock } = require('./helpers');
const webhook = require('../src/webhook');

jest.mock('../src/handler');
const handler = mocked(require('../src/handler'));

let newCtx = ctx;

beforeEach(() => (newCtx = { ...ctx }));

describe('webhook', () => {
  it('works with messages', () => {
    /** @type {import('serverless-telegram').Message} */
    // @ts-ignore
    const message = {
      message_id: 35,
      from: FROM,
      chat: CHAT,
      date: 1611414446,
      text: 'https://v.redd.it/rk2repuuvc261',
      entities: [],
    };
    const msgBotResponse = { video: '<file ID>', reply_to_message_id: 35 };
    /** @type {import('serverless-telegram').HttpRequest} */
    // @ts-ignore
    const msgHttpReq = { body: { update_id: 411306077, message } };
    const msgHttpRes = {
      body: { method: 'sendVideo', chat_id: CHAT.id, ...msgBotResponse },
      headers: { 'Content-Type': 'application/json' },
    };
    return withFnMocks(
      () => expect(webhook(newCtx, msgHttpReq)).resolves.toEqual(msgHttpRes),
      fnMock(handler.message, undefined, (msg) => {
        expect(msg).toEqual(message);
        return msgBotResponse;
      }),
    );
  });

  it('works with inline queries', () => {
    const inline_query = {
      id: '260980482508873432',
      from: FROM,
      query: 'https://v.redd.it/rk2repuuvc261',
      offset: '',
    };
    const inlineBotResponse = [
      { title: 'Send video (rk2repuuvc261.mp4)', video_file_id: '<file ID>' },
    ];
    /** @type {import('serverless-telegram').HttpRequest} */
    // @ts-ignore
    const inlineHttpReq = { body: { update_id: 411306081, inline_query } };
    const inlineHttpRes = {
      body: {
        method: 'answerInlineQuery',
        inline_query_id: '260980482508873432',
        results: [{ type: 'video', id: '0', ...inlineBotResponse[0] }],
      },
      headers: { 'Content-Type': 'application/json' },
    };
    return withFnMocks(
      () =>
        expect(webhook(newCtx, inlineHttpReq)).resolves.toEqual(inlineHttpRes),
      fnMock(handler.inline, undefined, (inline) => {
        expect(inline).toEqual(inline_query);
        return inlineBotResponse;
      }),
    );
  });

  it('ignores other updates', () => {
    /** @type {import('serverless-telegram').HttpRequest} */
    // @ts-ignore
    const emptyHttpReq = {};
    return expect(webhook(newCtx, emptyHttpReq)).resolves.toBeUndefined();
  });
});
