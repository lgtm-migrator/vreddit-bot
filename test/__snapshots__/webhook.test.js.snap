// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createForm creates a form for a video reply 1`] = `
FormData {
  "_boundary": "--------------------------555555555555555555555555",
  "_currentStream": null,
  "_insideLoop": false,
  "_overheadLength": 592,
  "_pendingNext": false,
  "_released": false,
  "_streams": Array [
    "----------------------------555555555555555555555555
Content-Disposition: form-data; name=\\"chat_id\\"

",
    "1",
    [Function],
    "----------------------------555555555555555555555555
Content-Disposition: form-data; name=\\"video\\"; filename=\\"hf352syjjka61.mp4\\"
Content-Type: video/mp4

",
    DelayedStream {
      "_bufferedEvents": Array [
        Arguments [
          "pause",
        ],
      ],
      "_events": Object {
        "error": [Function],
      },
      "_eventsCount": 1,
      "_maxDataSizeExceeded": false,
      "_released": false,
      "dataSize": 0,
      "maxDataSize": Infinity,
      "pauseStream": true,
      "source": ReadStream {
        "_events": Object {
          "data": [Function],
          "end": [Function],
          "error": [Function],
        },
        "_eventsCount": 3,
        "_maxListeners": undefined,
        "_readableState": ReadableState {
          "autoDestroy": false,
          "awaitDrainWriters": null,
          "buffer": BufferList {
            "head": null,
            "length": 0,
            "tail": null,
          },
          "closeEmitted": false,
          "closed": false,
          "decoder": null,
          "defaultEncoding": "utf8",
          "destroyed": false,
          "emitClose": true,
          "emittedReadable": false,
          "encoding": null,
          "endEmitted": false,
          "ended": false,
          "errorEmitted": false,
          "errored": false,
          "flowing": false,
          "highWaterMark": 65536,
          "length": 0,
          "multiAwaitDrain": false,
          "needReadable": false,
          "objectMode": false,
          "pipes": Array [],
          "readableListening": false,
          "reading": false,
          "readingMore": false,
          "resumeScheduled": false,
          "sync": true,
          Symbol(kPaused): true,
        },
        "autoClose": true,
        "bytesRead": 0,
        "closed": false,
        "emit": [Function],
        "end": Infinity,
        "fd": null,
        "flags": "r",
        "mode": 438,
        "path": "/tmp/hf352syjjka61.mp4",
        "pos": undefined,
        "start": undefined,
        Symbol(kFs): Object {
          "Dir": [Function],
          "Dirent": [Function],
          "F_OK": 0,
          "FileReadStream": [Function],
          "FileWriteStream": [Function],
          "R_OK": 4,
          "ReadStream": [Function],
          "Stats": [Function],
          "W_OK": 2,
          "WriteStream": [Function],
          "X_OK": 1,
          "_toUnixTimestamp": [Function],
          "access": [Function],
          "accessSync": [Function],
          "appendFile": [Function],
          "appendFileSync": [Function],
          "chmod": [Function],
          "chmodSync": [Function],
          "chown": [Function],
          "chownSync": [Function],
          "close": [Function],
          "closeSync": [Function],
          "constants": Object {
            "COPYFILE_EXCL": 1,
            "COPYFILE_FICLONE": 2,
            "COPYFILE_FICLONE_FORCE": 4,
            "F_OK": 0,
            "O_APPEND": 1024,
            "O_CREAT": 64,
            "O_DIRECT": 16384,
            "O_DIRECTORY": 65536,
            "O_DSYNC": 4096,
            "O_EXCL": 128,
            "O_NOATIME": 262144,
            "O_NOCTTY": 256,
            "O_NOFOLLOW": 131072,
            "O_NONBLOCK": 2048,
            "O_RDONLY": 0,
            "O_RDWR": 2,
            "O_SYNC": 1052672,
            "O_TRUNC": 512,
            "O_WRONLY": 1,
            "R_OK": 4,
            "S_IFBLK": 24576,
            "S_IFCHR": 8192,
            "S_IFDIR": 16384,
            "S_IFIFO": 4096,
            "S_IFLNK": 40960,
            "S_IFMT": 61440,
            "S_IFREG": 32768,
            "S_IFSOCK": 49152,
            "S_IRGRP": 32,
            "S_IROTH": 4,
            "S_IRUSR": 256,
            "S_IRWXG": 56,
            "S_IRWXO": 7,
            "S_IRWXU": 448,
            "S_IWGRP": 16,
            "S_IWOTH": 2,
            "S_IWUSR": 128,
            "S_IXGRP": 8,
            "S_IXOTH": 1,
            "S_IXUSR": 64,
            "UV_DIRENT_BLOCK": 7,
            "UV_DIRENT_CHAR": 6,
            "UV_DIRENT_DIR": 2,
            "UV_DIRENT_FIFO": 4,
            "UV_DIRENT_FILE": 1,
            "UV_DIRENT_LINK": 3,
            "UV_DIRENT_SOCKET": 5,
            "UV_DIRENT_UNKNOWN": 0,
            "UV_FS_COPYFILE_EXCL": 1,
            "UV_FS_COPYFILE_FICLONE": 2,
            "UV_FS_COPYFILE_FICLONE_FORCE": 4,
            "UV_FS_O_FILEMAP": 0,
            "UV_FS_SYMLINK_DIR": 1,
            "UV_FS_SYMLINK_JUNCTION": 2,
            "W_OK": 2,
            "X_OK": 1,
          },
          "copyFile": [Function],
          "copyFileSync": [Function],
          "createReadStream": [Function],
          "createWriteStream": [Function],
          "exists": [Function],
          "existsSync": [Function],
          "fchmod": [Function],
          "fchmodSync": [Function],
          "fchown": [Function],
          "fchownSync": [Function],
          "fdatasync": [Function],
          "fdatasyncSync": [Function],
          "fstat": [Function],
          "fstatSync": [Function],
          "fsync": [Function],
          "fsyncSync": [Function],
          "ftruncate": [Function],
          "ftruncateSync": [Function],
          "futimes": [Function],
          "futimesSync": [Function],
          "lchmod": undefined,
          "lchmodSync": undefined,
          "lchown": [Function],
          "lchownSync": [Function],
          "link": [Function],
          "linkSync": [Function],
          "lstat": [Function],
          "lstatSync": [Function],
          "lutimes": [Function],
          "lutimesSync": [Function],
          "mkdir": [Function],
          "mkdirSync": [Function],
          "mkdtemp": [Function],
          "mkdtempSync": [Function],
          "open": [Function],
          "openSync": [Function],
          "opendir": [Function],
          "opendirSync": [Function],
          "promises": Object {
            "access": [Function],
            "appendFile": [Function],
            "chmod": [Function],
            "chown": [Function],
            "copyFile": [Function],
            "lchmod": [Function],
            "lchown": [Function],
            "link": [Function],
            "lstat": [Function],
            "lutimes": [Function],
            "mkdir": [Function],
            "mkdtemp": [Function],
            "open": [Function],
            "opendir": [Function],
            "readFile": [Function],
            "readdir": [Function],
            "readlink": [Function],
            "realpath": [Function],
            "rename": [Function],
            "rmdir": [Function],
            "stat": [Function],
            "symlink": [Function],
            "truncate": [Function],
            "unlink": [Function],
            "utimes": [Function],
            "writeFile": [Function],
          },
          "read": [Function],
          "readFile": [Function],
          "readFileSync": [Function],
          "readSync": [Function],
          "readdir": [Function],
          "readdirSync": [Function],
          "readlink": [Function],
          "readlinkSync": [Function],
          "readv": [Function],
          "readvSync": [Function],
          "realpath": [Function],
          "realpathSync": [Function],
          "rename": [Function],
          "renameSync": [Function],
          "rmdir": [Function],
          "rmdirSync": [Function],
          "stat": [Function],
          "statSync": [Function],
          "symlink": [Function],
          "symlinkSync": [Function],
          "truncate": [Function],
          "truncateSync": [Function],
          "unlink": [Function],
          "unlinkSync": [Function],
          "unwatchFile": [Function],
          "utimes": [Function],
          "utimesSync": [Function],
          "watch": [Function],
          "watchFile": [Function],
          "write": [Function],
          "writeFile": [Function],
          "writeFileSync": [Function],
          "writeSync": [Function],
          "writev": [Function],
          "writevSync": [Function],
        },
        Symbol(kCapture): false,
        Symbol(kIsPerformingIO): false,
      },
    },
    [Function],
    "----------------------------555555555555555555555555
Content-Disposition: form-data; name=\\"width\\"

",
    "406",
    [Function],
    "----------------------------555555555555555555555555
Content-Disposition: form-data; name=\\"height\\"

",
    "720",
    [Function],
    "----------------------------555555555555555555555555
Content-Disposition: form-data; name=\\"reply_to_message_id\\"

",
    "2",
    [Function],
  ],
  "_valueLength": 3264572,
  "_valuesToMeasure": Array [],
  "dataSize": 0,
  "maxDataSize": 2097152,
  "pauseStreams": true,
  "readable": true,
  "writable": false,
}
`;
